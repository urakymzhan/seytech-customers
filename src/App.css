* {
  box-sizing: border-box;
}

/* .app {
  width: 200px;
  margin: 0 auto;
} */

/* Customers */
.customers img {
  width: 50px;
  border-radius: 50px;
}
.search-wrapper {
  /* border-bottom: solid 1px #aaa;
  padding: 5px;
  padding-left: 20px; */
  /* border: 1px solid red; */
  flex: 1;
}
.search-input,
.search-select {
  margin-bottom: 10px;
}

/* MENU */
.menu {
  list-style: none;
  display: flex;
  justify-content: space-between;
  padding: 10px 0;
  border-bottom: 1px solid gray;
}

/* .login-container {
  border: 1px solid red;
}
.customers-wrapper {
  border: 1px solid red;
} */
.customers-wrapper table {
  font-size: 12px;
}
.customers-wrapper table .btn {
  font-size: 12px;
  padding: 2px 4px;
}

/* single Customer page */
.single-customer-wrapper {
  border: 1px solid gray;
}
.single-customer-wrapper .row {
  border-bottom: 1px solid gray;
  display: flex;
}
.single-customer-wrapper .title {
  flex: 1;
  border-right: 1px solid gray;
  padding: 5px;
}
.single-customer-wrapper .desc {
  flex: 4;
  padding: 5px;
}
.single-customer-wrapper .desc input {
  width: 100%;
}
.single-customer-wrapper .btn {
  margin: 10px;
}
