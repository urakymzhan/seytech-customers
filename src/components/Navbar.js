import React from 'react';
import { isAuthenticated } from '../auth';
import { Link } from 'react-router-dom';

export default function Navbar({ logOut }) {
  return (
    <ul className="menu">
      {!isAuthenticated() && (
        <li>
          {' '}
          <Link to="/">Home</Link>{' '}
        </li>
      )}
      {!isAuthenticated() && (
        <li>
          {' '}
          <Link
            style={{
              background: '#007bff',
              color: '#fff',
              padding: '4px 6px',
              borderRadius: '4px',
            }}
            to="/login"
          >
            Login
          </Link>{' '}
        </li>
      )}
      {isAuthenticated() && (
        <li>
          {' '}
          <Link to="/dashboard">Dashboard</Link>{' '}
        </li>
      )}

      {/* todo: only available to super admin  */}
      {isAuthenticated() && (
        <li>
          {' '}
          <Link to="/customers">Customers</Link>{' '}
        </li>
      )}

      {!isAuthenticated() && (
        <li>
          {' '}
          <Link to="/about">About</Link>{' '}
        </li>
      )}
      <li>
        {' '}
        <Link to="/contact">Contact</Link>{' '}
      </li>

      {isAuthenticated() && (
        <li>
          <Link to="/login" onClick={logOut}>
            Logout
          </Link>
        </li>
      )}
    </ul>
  );
}
