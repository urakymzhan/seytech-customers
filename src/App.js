import React, { Component } from 'react';
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Link,
  Redirect,
} from 'react-router-dom';
import { Container } from 'reactstrap';
import Alert from 'reactstrap/lib/Alert';
import Customers from './components/Customers';
import SingleCustomer from './components/SingleCustomer';
import Login from './components/Login';
import './App.css';
import Home from './components/Home';
import About from './components/About';
import Contact from './components/Contact';
import NoMatch from './components/NoMatch';
import { isAuthenticated, removeToken } from './auth';
// import Cookies from 'js-cookie';
import { customersUrl, mainUrl, createurl } from './components/api';

class App extends Component {
  constructor() {
    super();
    this.state = {
      customerName: '',
      isLoading: false,
      customers: [],
      error: '',
      notification: '',
    };
  }

  getCustomers = () => {
    this.setState({ isLoading: true });
    fetch(customersUrl, {
      method: 'GET',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
    })
      .then((res) => {
        if (res.ok) {
          return res.json();
        } else {
          throw new Error('Server error! Please try again later.');
        }
      })
      .then((data) => {
        this.setState({
          customers: data.customers,
          isLoading: false,
        });
        setTimeout(() => {
          this.setState({ error: '' });
        }, 2500);
      })
      .catch((err) => {
        this.setState({ error: err.message });
      });
  };
  componentDidMount() {
    this.getCustomers();
  }
  delete = (customerId) => {
    fetch(`${mainUrl}/customer/${customerId}`, {
      method: 'delete',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
    })
      .then((res) => {
        if (res.ok) {
          return res.json();
        } else {
          throw new Error('Server error! Please try again later.');
        }
      })
      .then((data) => {
        const updatedCustomers = this.state.customers.filter(
          (customer) => customer._id !== customerId
        );
        this.setState({
          notification: data.message,
          customers: updatedCustomers,
        });
        setTimeout(() => {
          this.setState({ notification: '' });
        }, 2500);
      })
      .catch((err) => this.setState({ notification: err.message }));
  };

  logOut = () => {
    // in production
    // Cookies.remove('token');
    // simply rerender
    this.setState({ customerName: '' });
    // in development
    removeToken();
  };

  addCustomer = (customer) => {
    fetch(createurl, {
      method: 'post',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(customer),
    })
      .then((res) => {
        if (res.ok) {
          return res.json();
        } else {
          throw new Error('Server error! Please try again later.');
        }
      })
      .then((data) => {
        if (!data.customer) {
          throw new Error(data.message);
        }
        const updatedCustomers = [...this.state.customers, data.customer];
        console.log('updatedCustomers', updatedCustomers);
        this.setState({
          notification: data.message,
          customers: updatedCustomers,
        });
        setTimeout(() => {
          this.setState({ notification: '' });
        }, 2500);
      })
      .catch((err) => {
        this.setState({ notification: err.message });
        setTimeout(() => {
          this.setState({ notification: '' });
        }, 2500);
      });
  };

  onLoginSubmit = (customerName, token) => {
    // only in development
    localStorage.setItem('token', token);
    localStorage.setItem('customerName', customerName);
    // simply rerender
    this.setState({ customerName: customerName });
    this.getCustomers();
  };
  render() {
    const {
      customerName,
      isLoading,
      customers,
      error,
      notification,
    } = this.state;
    console.log('customers', customers);

    let customersContent;
    if (isLoading) {
      customersContent = <div>Loading...</div>;
    }
    if (error) {
      customersContent = (
        <Alert color="danger" fade={true}>
          {error}
        </Alert>
      );
    }
    if (customers.length === 0) {
      customersContent = isAuthenticated() ? (
        <Customers customers={[]} addCustomer={this.addCustomer} />
      ) : (
        <Redirect to="/login" />
      );
    }
    if (customers.length > 0) {
      customersContent = isAuthenticated() ? (
        <Customers
          addCustomer={this.addCustomer}
          delete={this.delete}
          customerName={customerName}
          getCustomers={this.getCustomers}
          customers={customers}
          notification={notification}
        />
      ) : (
        <Redirect to="/login" />
      );
    }

    return (
      <Container fluid="xl">
        <Router>
          <ul className="menu">
            <li>
              {' '}
              <Link to="/">Home</Link>{' '}
            </li>
            <li>
              {' '}
              <Link to="/about">About</Link>{' '}
            </li>
            <li>
              {' '}
              <Link to="/contact">Contact</Link>{' '}
            </li>
            <li>
              {' '}
              <Link to="/customers">Customers</Link>{' '}
            </li>
            {isAuthenticated() && (
              <li>
                <Link to="/login" onClick={this.logOut}>
                  Logout
                </Link>
              </li>
            )}
          </ul>

          <Switch>
            <Route exact path="/">
              <Home />
            </Route>
            <Route exact path="/about">
              <About />
            </Route>
            <Route exact path="/contact">
              <Contact />
            </Route>

            <Route exact path="/customers">
              {customersContent}
            </Route>
            <Route path="/customer/:id">
              <SingleCustomer delete={this.delete} customers={customers} />
            </Route>
            <Route path="/customer/:id/:action">
              <SingleCustomer delete={this.delete} customers={customers} />
            </Route>
            <Route exact path="/login">
              {isAuthenticated() ? (
                <Redirect to="/customers" />
              ) : (
                // <div>Already logged in</div>
                <Login onLoginSubmit={this.onLoginSubmit} />
              )}
            </Route>
            <Route path="*">
              <NoMatch />
            </Route>
          </Switch>
        </Router>
      </Container>
    );
  }
}

export default App;
