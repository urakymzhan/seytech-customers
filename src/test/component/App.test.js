import React from 'react';
// import { render } from '@testing-library/react';
import App from '../../App';
import Navbar from '../../components/Navbar';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

describe('<App> component tests', () => {
  it('renders <App> component correctly', () => {
    shallow(<App />);
  });
  test('snapshot renders', () => {
    const component = renderer.create(<App />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot(); // https://jestjs.io/docs/en/snapshot-testing
  });

  it('renders <Navbar> component', () => {
    const wrapper = shallow(<App />);
    expect(wrapper.find(Navbar).length).toEqual(1);
  });

  // ask them to add more tests...

  it('renders 1 menu items correctly', () => {
    let app = mount(<App />);
    // console.log(app.debug());
    expect(app.find('.menu').length).toEqual(1);
  });
});
