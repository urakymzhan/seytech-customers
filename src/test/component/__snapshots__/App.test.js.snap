// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App> component tests snapshot renders 1`] = `
Array [
  <ul
    className="menu"
  >
    <li>
       
      <a
        href="/"
        onClick={[Function]}
      >
        Home
      </a>
       
    </li>
    <li>
       
      <a
        href="/login"
        onClick={[Function]}
        style={
          Object {
            "background": "#007bff",
            "borderRadius": "4px",
            "color": "#fff",
            "padding": "4px 6px",
          }
        }
      >
        Login
      </a>
       
    </li>
    <li>
       
      <a
        href="/about"
        onClick={[Function]}
      >
        About
      </a>
       
    </li>
    <li>
       
      <a
        href="/contact"
        onClick={[Function]}
      >
        Contact
      </a>
       
    </li>
  </ul>,
  <div
    className="container-xl"
    style={
      Object {
        "minHeight": "100vh",
        "padding": "10px",
      }
    }
  >
    <div>
      <h1>
        Seytech
      </h1>
      <p>
        Fullstack Software Developer Bootcamp
      </p>
      <h5>
        WHERE OUR GRADS WORK
      </h5>
      <h5>
        Testimonials
      </h5>
      <h5>
        AVAILABLE COURSES
      </h5>
      <h5>
        Full-Time! Real Time! Online!
      </h5>
      <h5>
        ACT.JS What You Will Learn!
      </h5>
      <h5>
        Why Choose Use?
      </h5>
      <h5>
        Our Team
      </h5>
      <h5>
        AREER CAREER SERVICES
      </h5>
      <h5>
        Our Alumni
      </h5>
      <h5>
        Placement-Based Tuition
      </h5>
      <h5>
        Get In Touch
      </h5>
    </div>
  </div>,
]
`;
